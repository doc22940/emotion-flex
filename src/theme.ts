export interface Theme {
  breakpoints: {
    sm: number;
    md: number;
    lg: number;
    xl: number;
  };
  maxWidths: {
    sm: string;
    md: string;
    lg: string;
    xl: string;
  };
  gutters: number;
}

const defaultTheme = {
  breakpoints: {
    sm: 576,
    md: 768,
    lg: 992,
    xl: 1200,
  },
  maxWidths: {
    sm: '540px',
    md: '720px',
    lg: '960px',
    xl: '1140px',
  },
  gutters: 15,
};

export default defaultTheme;
