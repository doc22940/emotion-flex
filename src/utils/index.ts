import defaultTheme, { Theme } from '../theme';

// https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Number/isInteger#Polyfill
export const isInteger = (value: number) =>
  typeof value === 'number' && isFinite(value) && Math.floor(value) === value;

export const getBreakpoints = (theme: Theme) =>
  Object.values(theme.breakpoints).map(bp => `@media (min-width: ${bp}px)`);

export const getTheme = (theme: Theme | {}) => {
  const themeConfig = theme || {};

  return {
    ...defaultTheme,
    // @ts-ignore
    breakpoints: { ...defaultTheme.breakpoints, ...themeConfig.breakpoints },
    // @ts-ignore
    maxWidths: { ...defaultTheme.maxWidths, ...themeConfig.maxWidths },
    ...themeConfig,
  };
};
