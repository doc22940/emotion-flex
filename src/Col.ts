import facepaint from 'facepaint';
import styled from './styled';
import { isInteger, getTheme, getBreakpoints } from './utils';

type TwoToEleven = 2 | 3 | 4 | 5 | 6 | 7 | 8 | 9 | 10 | 11;
type OneToTwelve = 1 | TwoToEleven | 12;
type ZeroToTwelve = 0 | OneToTwelve;

type Offset = 0 | 1 | TwoToEleven;
type Order = ZeroToTwelve | 'first' | 'last';
type Column = OneToTwelve;

export interface ColProps {
  noGutters?: boolean;
  xsOffset?: Offset;
  xsOrder?: Order;
  xs?: Column;
  smOffset?: Offset;
  smOrder?: Order;
  sm?: Column;
  mdOffset?: Offset;
  mdOrder?: Order;
  md?: Column;
  lgOffset?: Offset;
  lgOrder?: Order;
  lg?: Column;
  xlOffset?: Offset;
  xlOrder?: Order;
  xl?: Column;
}

const toOrder = (order?: Order) => {
  if (!order || typeof order === 'number') {
    return order;
  }
  return order === 'first' ? -1 : 13;
};

const toPercent = (n: number) => {
  const percentage = (n / 12) * 100;
  return isInteger(percentage) ? `${percentage}%` : `${percentage.toFixed(6)}%`;
};

export const Col = styled.div<ColProps>`
  position: relative;
  width: 100%;
  padding-right: ${({ noGutters, theme }) =>
    noGutters ? 0 : getTheme(theme).gutters}px;
  padding-left: ${({ noGutters, theme }) =>
    noGutters ? 0 : getTheme(theme).gutters}px;

  ${p =>
    facepaint(getBreakpoints(getTheme(p.theme)))({
      marginLeft: [
        p.xsOffset && toPercent(p.xsOffset),
        p.smOffset && toPercent(p.smOffset),
        p.mdOffset && toPercent(p.mdOffset),
        p.lgOffset && toPercent(p.lgOffset),
        p.xlOffset && toPercent(p.xlOffset),
      ],
      flex: [
        p.xs ? `1 0 ${toPercent(p.xs)}` : '1 0 0',
        p.sm && `1 0 ${toPercent(p.sm)}`,
        p.md && `1 0 ${toPercent(p.md)}`,
        p.lg && `1 0 ${toPercent(p.lg)}`,
        p.xl && `1 0 ${toPercent(p.xl)}`,
      ],
      maxWidth: [
        p.xs ? toPercent(p.xs) : '100%',
        p.sm && toPercent(p.sm),
        p.md && toPercent(p.md),
        p.lg && toPercent(p.lg),
        p.xl && toPercent(p.xl),
      ],
      order: [
        toOrder(p.xsOrder),
        toOrder(p.smOrder),
        toOrder(p.mdOrder),
        toOrder(p.lgOrder),
        toOrder(p.xlOrder),
      ],
    })}
`;

Col.displayName = 'Col';
