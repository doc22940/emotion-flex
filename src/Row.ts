import styled from './styled';
import { getTheme } from './utils';

type Align = 'start' | 'center' | 'end';
type Justify = 'start' | 'center' | 'end' | 'between' | 'around';

export interface RowProps {
  noGutters?: boolean;
  alignItems?: Align; // TODO: add align-items
  justifyContent?: Justify; // TODO: add justify-content
}

export const Row = styled.div<RowProps>`
  display: flex;
  flex-wrap: wrap;
  margin-left: ${({ noGutters, theme }) =>
    noGutters ? 0 : `-${getTheme(theme).gutters}px`};
  margin-right: ${({ noGutters, theme }) =>
    noGutters ? 0 : `-${getTheme(theme).gutters}px`};
`;

Row.displayName = 'Row';
