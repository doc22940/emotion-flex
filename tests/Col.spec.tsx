import React from 'react';
import { create } from 'react-test-renderer';

import { Col } from '../src';

describe('<Col />', () => {
  it('renders with default styles', () => {
    const tree = create(<Col />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it('renders with responsive styles', () => {
    const tree = create(<Col xs={12} sm={6} md={4} lg={3} xl={1} />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
