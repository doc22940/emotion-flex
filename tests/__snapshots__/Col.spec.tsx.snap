// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Col /> renders with default styles 1`] = `
.emotion-0 {
  position: relative;
  width: 100%;
  padding-right: 15px;
  padding-left: 15px;
  -webkit-flex: 1 0 0;
  -ms-flex: 1 0 0;
  flex: 1 0 0;
  max-width: 100%;
}

<div
  className="emotion-0"
/>
`;

exports[`<Col /> renders with responsive styles 1`] = `
.emotion-0 {
  position: relative;
  width: 100%;
  padding-right: 15px;
  padding-left: 15px;
  -webkit-flex: 1 0 100%;
  -ms-flex: 1 0 100%;
  flex: 1 0 100%;
  max-width: 100%;
}

@media (min-width:576px) {
  .emotion-0 {
    -webkit-flex: 1 0 50%;
    -ms-flex: 1 0 50%;
    flex: 1 0 50%;
    max-width: 50%;
  }
}

@media (min-width:768px) {
  .emotion-0 {
    -webkit-flex: 1 0 33.333333%;
    -ms-flex: 1 0 33.333333%;
    flex: 1 0 33.333333%;
    max-width: 33.333333%;
  }
}

@media (min-width:992px) {
  .emotion-0 {
    -webkit-flex: 1 0 25%;
    -ms-flex: 1 0 25%;
    flex: 1 0 25%;
    max-width: 25%;
  }
}

@media (min-width:1200px) {
  .emotion-0 {
    -webkit-flex: 1 0 8.333333%;
    -ms-flex: 1 0 8.333333%;
    flex: 1 0 8.333333%;
    max-width: 8.333333%;
  }
}

<div
  className="emotion-0"
/>
`;
